File: my-folder/1635-number-of-good-pairs/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var numIdenticalPairs = function(nums) {
    
    let f = x => x<=1 ? 1 : x*f(x-1)
    let comb = (n, r) => f(n) / (f(r) * f(n-r))
    
    let map = new Map()
    
    for (let num of nums) {
        if (!map.has(num)) {
            map.set(num, 1)
        }
        else {
            map.set(num, map.get(num)+1)
        }
    }
    
    let totalCount = 0
    for (entry of map.entries()) {
        if (entry[1] > 1) {
            totalCount += comb(entry[1], 2)
        }
    }
    
    return totalCount
    
};

File: my-folder/1635-number-of-good-pairs/solution.py
class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        
        @cache
        def f(n):
            if n<=1: return 1
            return n * f(n-1)
        
        def comb(n, r):
            return f(n) // (f(r) * f(n-r))
        
        numCounts = Counter(nums)
        
        totalCount = 0
        
        for num, count in numCounts.items():
            if count > 1:
                totalCount += comb(count, 2)
            
        return totalCount
